app/src/main/java/org/schabi/newpipe/player/MainPlayer.java

```diff
@@ -0,0 +1,379 @@
+    }
+
+    @Override
+    public int onStartCommand(Intent intent, int flags, int startId) {
+        if (DEBUG) Log.d(TAG, "onStartCommand() called with: intent = [" + intent +
+                "], flags = [" + flags + "], startId = [" + startId + "]");
+
+        if (Intent.ACTION_MEDIA_BUTTON.equals(intent.getAction()) || intent.getStringExtra(VideoPlayer.PLAY_QUEUE_KEY) != null)
+            showNotificationAndStartForeground();
+
+        playerImpl.handleIntent(intent);
+        if (playerImpl.mediaSessionManager != null) {
+            playerImpl.mediaSessionManager.handleMediaButtonIntent(intent);
+        }
+        return START_NOT_STICKY;
+    }
+
+    public void stop(boolean autoplayEnabled) {
+        if (DEBUG) Log.d(TAG, "stop() called");
+
+        if (playerImpl.getPlayer() != null) {
+            playerImpl.wasPlaying = playerImpl.getPlayer().getPlayWhenReady();
+            // Releases wifi & cpu, disables keepScreenOn, etc.
+            if (!autoplayEnabled) playerImpl.onPause();
+            // We can't just pause the player here because it will make transition from one stream to a new stream not smooth
+            playerImpl.getPlayer().stop(false);
+            playerImpl.setRecovery();
+            // Notification shows information about old stream but if a user selects a stream from backStack it's not actual anymore
+            // So we should hide the notification at all.
+            // When autoplay enabled such notification flashing is annoying so skip this case
+            if (!autoplayEnabled) stopForeground(true);
+        }
+    }
+
+    @Override
+    public void onTaskRemoved(Intent rootIntent) {
+        super.onTaskRemoved(rootIntent);
+        onDestroy();
+        // Unload from memory completely
+        Runtime.getRuntime().halt(0);
+    }
+
+    @Override
+    public void onDestroy() {
+        if (DEBUG) Log.d(TAG, "destroy() called");
+        onClose();
+    }
+
+    @Override
+    protected void attachBaseContext(Context base) {
+        super.attachBaseContext(AudioServiceLeakFix.preventLeakOf(base));
+    }
+
+    @Override
+    public IBinder onBind(Intent intent) {
+        return mBinder;
+    }
+
+    /*//////////////////////////////////////////////////////////////////////////
+    // Actions
+    //////////////////////////////////////////////////////////////////////////*/
+    private void onClose() {
+        if (DEBUG) Log.d(TAG, "onClose() called");
+
+        if (lockManager != null) {
+            lockManager.releaseWifiAndCpu();
+        }
+        if (playerImpl != null) {
+            removeViewFromParent();
+
+            playerImpl.setRecovery();
+            playerImpl.savePlaybackState();
+            playerImpl.stopActivityBinding();
+            playerImpl.removePopupFromView();
+            playerImpl.destroy();
+        }
+        if (notificationManager != null) notificationManager.cancel(NOTIFICATION_ID);
+        playerImpl = null;
+        lockManager = null;
```

Comment:
Since in that place the service is stopping, no need all those `something = null` because they will be cleared from memory anyway, so I just removed these assigments